feat: add support for DNS

adeatcu-ionos opened the pull request:
## What does this fix or implement?

<!-- Enter details of the change here. Include additional tests that have been done, reference to the issue for tracking, etc. -->

## Checklist

<!-- Please check the completed items below -->
<!-- Not all changes require documentation updates or tests to be added or updated -->

- [x] PR name added as appropriate (e.g. `feat:`/`fix:`/`doc:`/`test:`/`refactor:`)
- [ ] Tests added or updated
- [ ] Documentation updated
- [ ] Changelog updated and version incremented (label: upcoming release)
- [ ] Github Issue linked if any
- [ ] Jira task updated

adeatcu-ionos commented:
@cristiGuranIonos please check this out when you have time, this is the first iteration. During your review, I'll write the documentation and after this initial review, I'll also write the tests. Thank you!

cristiGuranIonos commented on services/dnsaas/zone.go:
```diff
@@ -0,0 +1,142 @@
+	if zone.Id != nil {
+		d.SetId(*zone.Id)
+	}
+
```
if zone.properties == nil - return error

adeatcu-ionos replied:
@cristiGuranIonos solved in 2273efbf55a163a4f7e5e4a2f4f6bbe854780969 for both zones and records. I also added documentation. If everything else is ok, I can start writing tests.

cristiGuranIonos commented on ionoscloud/data_source_dnsaas_record.go:
```diff
@@ -0,0 +1,143 @@
+	return &schema.Resource{
+		ReadContext: dataSourceRecordRead,
+		Schema: map[string]*schema.Schema{
+			"id": {
```
maybe a UUID validation here and for zone_id

adeatcu-ionos replied:
@cristiGuranIonos done in 01e2b65f9fe5d96e6c21367e7fe3b2e7b6343ba2

cristiGuranIonos commented on ionoscloud/data_source_dnsaas_record.go:
```diff
@@ -0,0 +1,143 @@
+			if err != nil {
+				return diag.FromErr(fmt.Errorf("an error occured while fetching DNS Records: %w", err))
+			}
+			if records.Items != nil {
```
maybe just break if records.Items == nil and the error will be handled by `if results == nil || len(results) == 0 {`

adeatcu-ionos replied:
@cristiGuranIonos done in 01e2b65f9fe5d96e6c21367e7fe3b2e7b6343ba2, I completely removed it.

cristiGuranIonos commented on ionoscloud/resource_dnsaas_record.go:
```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("an error occured while creating a record for the DNS zone with ID: %s, error: %w", zoneId, err))
+	}
+	if recordResponse.Metadata == nil {
```
Do we have to wait  until it's in `created` state? Or is it synchronous?

adeatcu-ionos commented on ionoscloud/data_source_dnsaas_record.go:
```diff
@@ -0,0 +1,143 @@
+			if err != nil {
+				return diag.FromErr(fmt.Errorf("an error occured while fetching DNS Records: %w", err))
+			}
+			if records.Items != nil {
```
@cristiGuranIonos done in 01e2b65f9fe5d96e6c21367e7fe3b2e7b6343ba2, I completely removed it.

adeatcu-ionos commented on ionoscloud/data_source_dnsaas_zone.go:
```diff
@@ -0,0 +1,133 @@
+			if err != nil {
+				return diag.FromErr(fmt.Errorf("an error occured while fetching DNS Zones: %w", err))
+			}
+			if zones.Items != nil {
```
@cristiGuranIonos done in 01e2b65f9fe5d96e6c21367e7fe3b2e7b6343ba2, I completely removed it.

cristiGuranIonos commented:
Add to changelog  please

adeatcu-ionos commented on ionoscloud/resource_dnsaas_record_test.go:
```diff
@@ -0,0 +1,218 @@
+package ionoscloud
```
@cristiGuranIonos done in 8689a85494ef2d919fd4174d649982cce2919387

adeatcu-ionos commented:
> Add to changelog please

@cristiGuranIonos done in 8689a85494ef2d919fd4174d649982cce2919387

mimihalescu commented on docs/data-sources/dnsaas_record.md:
```diff
@@ -0,0 +1,66 @@
+  Get information on a DNS Record.
+---
+
+# ionoscloud_dns_zone
```
ionoscloud_dns_record*

adeatcu-ionos replied:
@mimihalescu done, thanks for catching this!

adeatcu-ionos commented:
@mimihalescu @cristiGuranIonos I also made the renaming, please take a look and check if I missed something, thank you!

adeatcu-ionos commented on services/dns/record.go:
```diff
@@ -0,0 +1,209 @@
+func (c *Client) CreateRecord(ctx context.Context, zoneId string, d *schema.ResourceData) (recordResponse dns.RecordResponse, responseInfo utils.ApiResponseInfo, err error) {
+	// Must is a helper that wraps a call to a function returning (UUID, error) and panics if the error is non-nil.
+	uuidV4 := uuid.Must(uuid.NewV4())
+	recordUuid := uuid.NewV5(uuid.NewV5(uuid.NamespaceURL, "https://github.com/ionos-cloud/terraform-provider-ionoscloud"), uuidV4.String())
```
@cristiGuranIonos I had in mind to generate the record `UUID` also based on zone `UUID`, not sure if this adds anything useful tho, what do you think?

cristiGuranIonos replied:
I don't think it's necessary

adeatcu-ionos commented:
@mimihalescu @avirtopeanu-ionos @rmocanu-ionos @cristiGuranIonos I added all of you for the review since I changed the functionality for create method, so now the creation is done via `PUT` instead of `POST`, as we discussed, please take a look and tell me what you think.

cristiGuranIonos commented on go.mod:
```diff
@@ -3,6 +3,7 @@ module github.com/ionos-cloud/terraform-provider-ionoscloud/v6
 go 1.20
 
 require (
+	github.com/gofrs/uuid v4.4.0+incompatible
```
can we use v5 instead?

adeatcu-ionos replied:
@cristiGuranIonos done in 5369bab15a6a1015bc488c1140a42af0283ed025.

cristiGuranIonos commented on services/dns/record.go:
```diff
@@ -14,8 +15,11 @@ import (
-	recordResponse, apiResponse, err := c.sdkClient.RecordsApi.ZonesRecordsPost(ctx, zoneId).RecordCreateRequest(*request).Execute()
+	// Must is a helper that wraps a call to a function returning (UUID, error) and panics if the error is non-nil.
+	uuidV4 := uuid.Must(uuid.NewV4())
+	recordUuid := uuid.NewV5(uuid.NewV5(uuid.NamespaceURL, "https://github.com/ionos-cloud/terraform-provider-ionoscloud"), uuidV4.String())
```
I think we can extract this part as a constant? - uuid.NewV5(uuid.NamespaceURL, "https://github.com/ionos-cloud/terraform-provider-ionoscloud")

avirtopeanu-ionos commented on services/dns/zone.go:
```diff
@@ -14,8 +15,11 @@ import (
-	responseData, apiResponse, err := c.sdkClient.ZonesApi.ZonesPost(ctx).ZoneCreateRequest(*request).Execute()
+	// Must is a helper that wraps a call to a function returning (UUID, error) and panics if the error is non-nil.
+	uuidV4 := uuid.Must(uuid.NewV4())
+	zoneUuid := uuid.NewV5(uuid.NewV5(uuid.NamespaceURL, "https://github.com/ionos-cloud/terraform-provider-ionoscloud"), uuidV4.String())
```
in my opinion writing `uuid.NewV5(uuid.NewV5(uuid.NamespaceURL, "https://github.com/ionos-cloud/terraform-provider-ionoscloud"), uuidV4.String())` every time you want a new ID is a bit repetitive

What do you feel about creating a very small pkg `internal/tfid/tfid.go` which has the sole responsibility of just creating V5 UUIDs with the namespaceURL already set ? And you could call it like `tfid.New()` for example

adeatcu-ionos replied:
@avirtopeanu-ionos sure, not sure why I didn't do that in the beginning.

adeatcu-ionos replied:
@avirtopeanu-ionos done in 5369bab15a6a1015bc488c1140a42af0283ed025.

avirtopeanu-ionos commented on vendor/modules.txt:
```diff
@@ -9,6 +9,9 @@ github.com/apparentlymart/go-textseg/v13/textseg
 github.com/fatih/color
 # github.com/go-test/deep v1.0.6
 ## explicit; go 1.13
+# github.com/gofrs/uuid v4.4.0+incompatible
```
Any reason for using this lib ?

Google/uuid package is already in this go.mod and having 2 dependencies on different uuid pkgs isnt good

And the reason I prefer keeping `google/uuid` over refactoring to only use the new one is because any dev which just quickly wants an UUID V4 would default to `google/uuid` without thinking that we're actually using `gofrs/uuid`

avirtopeanu-ionos replied:
I looked over docs and seems like `google/uuid` cant generate V5 UUIDs.

In this case I'm okay with having two different libs for UUID handling, but we should get rid of `google/uuid` as soon as possible and replace all those calls with `gofrs/uuid` equivalent

adeatcu-ionos replied:
@avirtopeanu-ionos ` but we should get rid of google/uuid as soon as possible and replace all those calls with gofrs/uuid equivalent` I agree, I'll open an issue for this.

avirtopeanu-ionos commented on ionoscloud/resource_dns_zone.go:
```diff
@@ -0,0 +1,147 @@
+	if zoneResponse.Metadata.State != nil {
+		if *zoneResponse.Metadata.State == dns.FAILED {
+			// This is a temporary error message since right now the API is not returning errors that we can work with.
+			return diag.FromErr(fmt.Errorf("zone update has failed, this can happen if the data in the request is not correct, " +
```
In this case, the API zone is still going to remain in FAILED state, right? Does throwing an error not kind of hide the fact that something on the cloud has changed after running the plan? I'm not experienced with terraform, just trying to understand what is happening here

adeatcu-ionos replied:
`I'm not experienced with terraform` no worries

I understand your point and it makes sense, it can actually be an edge case that we I didn't think about, I will check to see the current behavior. I expect something like:

1) We try to create the Zone, we send the request and the zone is created in a `FAILED` state;
2) Because of that, we raise an error, and we are not setting anything in the `Terraform` state, this means that Terraform will not track anything related to this resource;
3) The resource will continue to exist in a `FAILED` state and basically this is not good since Terraform won't track this resource.

But I'll come back with updates.

avirtopeanu-ionos replied:
Ok so I did a bit more digging and (please correct any of my points if wrong):
- This is known as "terraform drift" and should be corrected externally (via CLI for example)
- Throwing an error in this case is good, since the terraform plan wasn't successful
- maybe consider updating the error message to let the user know:
     1. that a zone has been created/updated/etc., however it is in state `FAILED`
     2. he needs to take some external action to revert the tf plan's changes

adeatcu-ionos replied:
@avirtopeanu-ionos this is not really a drift since we are not setting anything in the terraform state. Usually, after you receive a response, you set some values in the state in order to have a mapping between the resource defined in the plan and the actual resurce that exists in the cloud. Also, you can check this:
```
Drift is the term for when the real-world state of your infrastructure differs from the state defined in your configuration. 
```
`Throwing an error in this case is good, since the terraform plan wasn't successful` indeed it is, but my problem is that the resource still exists in cloud, at least this is what I expect.
`maybe consider updating the error message to let the user know: ` this is a possible solution, we will check all possible solutions after I test the behavior. I can't test it RN because there are some problems with the DNS API.

adeatcu-ionos replied:
@avirtopeanu-ionos I was able to test this again, but it seems like the behavior is different, I tried to create a record in the `FAILED` state, but right now I receive a 400 error response, it seems like they changed the behavior. Let's leave it as it is for now and I will ask them how I can generate a response in which the state is `FAILED` in order to check if a resource is created or not.

avirtopeanu-ionos commented on ionoscloud/resource_dns_zone.go:
```diff
@@ -0,0 +1,147 @@
+
+	err = utils.WaitForResourceToBeDeleted(ctx, d, client.IsZoneDeleted)
+	if err != nil {
+		diag.FromErr(fmt.Errorf("an error occured while waiting for the DNS Zone with ID: %s to be deleted, error: %w", zoneId, err))
```
shouldn't `diag.FromErr` always be returned in case of err?

adeatcu-ionos replied:
@avirtopeanu-ionos done, I missed this.

avirtopeanu-ionos commented on services/dns/record.go:
```diff
@@ -0,0 +1,205 @@
+	}
+
+	if record.Properties.Name != nil {
+		if err := d.Set("name", *record.Properties.Name); err != nil {
+			return utils.GenerateSetError(recordResourceName, "name", err)
+		}
+	}
+
+	if record.Properties.Type != nil {
+		if err := d.Set("type", *record.Properties.Type); err != nil {
+			return utils.GenerateSetError(recordResourceName, "type", err)
+		}
+	}
+
+	if record.Properties.Content != nil {
+		if err := d.Set("content", *record.Properties.Content); err != nil {
+			return utils.GenerateSetError(recordResourceName, "content", err)
+		}
+	}
+
+	if record.Properties.Ttl != nil {
+		if err := d.Set("ttl", *record.Properties.Ttl); err != nil {
+			return utils.GenerateSetError(recordResourceName, "ttl", err)
+		}
+	}
+
+	if record.Properties.Enabled != nil {
+		if err := d.Set("enabled", *record.Properties.Enabled); err != nil {
+			return utils.GenerateSetError(recordResourceName, "enabled", err)
+		}
+	}
+
+	if record.Metadata.Fqdn != nil {
+		if err := d.Set("fqdn", *record.Metadata.Fqdn); err != nil {
+			return utils.GenerateSetError(recordResourceName, "fqdn", err)
+		}
+	}
+
+	return nil
+}
```
Consider github.com/fatih/structs for dynamic field estting
